File: 084_Largest_Rectangle_in_Histogram/084_1.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        int res = 0;
        int len = heights.size();
        stack<int> stack;
        for (int i = 0; i < len; i++) {
            while (!stack.empty() && heights[stack.top()] > heights[i]) {
                int curHeight = heights[stack.top()];
                stack.pop();
                while (!stack.empty() && heights[stack.top()] == curHeight)
                    stack.pop();
                int curWidth;
                if (stack.empty())
                    curWidth = i;
                else
                    curWidth = i - stack.top() - 1;
                res = max(res, curHeight * curWidth);
            }
            stack.push(i);
        }

        while (!stack.empty()) {
            int curHeight = heights[stack.top()];
            stack.pop();
            while (!stack.empty() && heights[stack.top()] == curHeight)
                stack.pop();
            int curWidth;
            if (stack.empty())
                curWidth = len;
            else
                curWidth = len - stack.top() - 1;
            res = max(res, curHeight * curWidth);
        }
        return res;
    }
};

File: 084_Largest_Rectangle_in_Histogram/084_1.py
from typing import List

class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        res = 0
        heights = [0] + heights + [0]
        stack = [0]
        for i in range(1, len(heights)):
            while heights[stack[-1]] > heights[i]:
                while stack[-1] == stack[-2]:
                    stack.pop()
                height = heights[stack.pop()]
                width = i - stack[-1] - 1
                res = max(res, height * width)
            stack.append(i)
        return res